### ISSUE TITLE
Internal error: encountered unplanned aggregate function

### ISSUE THREAD

**Issue by philip-stoev** [MaterializeInc/materialize#20902]
### What version of Materialize are you using?

 v0.61.0 (31314ef67)

### What is the issue?

```
DROP TABLE t1;
CREATE TABLE t1 (f1 INTEGER);

(
  SELECT 1
  FROM t1
)
UNION SELECT 1 AS cnt
  FROM t1
  ORDER BY SUM(f1) DESC
; 
```
Results in :

```
ERROR:  Internal error: encountered unplanned aggregate function: Item { id: System(244), qualifiers: ItemQualifiers { database_spec: Ambient, schema_spec: Id(System(2)) }, full_name: FullItemName { database: Ambient, schema: "pg_catalog", item: "sum" }, print_id: false }
```

Putting the second `SELECT` inside parentheses causes the issue to go away, which is concerning -- maybe we are not parsing UNION branches properly?

**Comment by antiguru** [comment]
I think this is not a compute problem but still within the SQL planner. Assigning to the Surfaces team.

**Comment by ggevay** [comment]
I think putting the second SELECT inside parentheses changes the meaning of the query:
- With parenthesis, it's clear that the ORDER BY belongs to just the second UNION input. (Which means that it won't actually do much, because the UNION is then free to change the order.)
- Without parenthesis, the ORDER BY concerns the final result of the query. Note that whether the ORDER BY concerns the final result of the query or still concerns the UNION input is a question of precedence: Does UNION or ORDER BY has higher precedence? Postgres says it's the UNION (`f1` would exist inside the UNION input, but not afterwards):
```
gabor=# DROP TABLE t1;
CREATE TABLE t1 (f1 INTEGER);

(
  SELECT 1
  FROM t1
)
UNION SELECT 1 AS cnt
  FROM t1
  ORDER BY SUM(f1) DESC
; 
DROP TABLE
CREATE TABLE
ERROR:  column "f1" does not exist
LINE 7:   ORDER BY SUM(f1) DESC
                       ^
```
Our parser thinks the same (which we can see if we write `f1` instead of `SUM(f1)`):
```
materialize=> DROP TABLE t1;
CREATE TABLE t1 (f1 INTEGER);

(
  SELECT 1
  FROM t1
)
UNION SELECT 1 AS cnt
  FROM t1
  ORDER BY f1 DESC
;
DROP TABLE
CREATE TABLE
ERROR:  column "f1" does not exist
```

But the bug is valid: we shouldn't be throwing an internal error in any case.

**Comment by ggevay** [comment]
Or wait, did you mean putting the parenthesis like this:
```
(
  SELECT 1
  FROM t1
)
UNION (SELECT 1 AS cnt
  FROM t1
  ORDER BY SUM(f1) DESC
)
;
```
or like this:
```
(
  SELECT 1
  FROM t1
)
UNION (SELECT 1 AS cnt
  FROM t1
)
ORDER BY SUM(f1) DESC
;
```
The latter makes the parsing more clear, but the former changes the meaning.

**Comment by ggevay** [comment]
Btw. Postgres doesn't support an aggregate in the ORDER BY of a set expression:
```
gabor=# (
  SELECT 1 as cnt
  FROM t1
)
UNION (SELECT 1 AS cnt
  FROM t1
)
ORDER BY SUM(cnt) DESC
;
ERROR:  invalid UNION/INTERSECT/EXCEPT ORDER BY clause
LINE 8: ORDER BY SUM(cnt) DESC
                 ^
DETAIL:  Only result column names can be used, not expressions or functions.
HINT:  Add the expression/function to every SELECT, or move the UNION into a FROM clause.
```
So then if we are ok with following Postgres in this, then the fix is as simple as flipping `allow_aggregates` to false [here](https://github.com/MaterializeInc/materialize/blob/d969cb534961302829210c654bfafa24809ebb73/src/sql/src/plan/query.rs#L1251), and clarifying in the scope `name` that it's not simply a `ORDER BY clause`, but `ORDER BY clause of a set expression`. And then we have the following error msg:
```
materialize=> CREATE TABLE t1 (f1 INTEGER);
CREATE TABLE
materialize=> (                            
  SELECT 1 as cnt
  FROM t1
)
UNION (SELECT 1 AS cnt
  FROM t1
)
ORDER BY SUM(cnt) DESC
;
ERROR:  aggregate functions are not allowed in ORDER BY clause of a set expression (function pg_catalog.sum)
```

**Comment by ggevay** [comment]
So the bug is that normally, aggregate functions in an ORDER BY clause are planned inside `plan_view_select` in Step 3 based on what `plan_query_inner` gives `plan_view_select` as the `order_by_exprs` argument, but set expressions are special: [Their ORDER BY is planned outside of `plan_view_select`](https://github.com/MaterializeInc/materialize/blob/d969cb534961302829210c654bfafa24809ebb73/src/sql/src/plan/query.rs#L1257), and thus the aggregate planning logic inside `plan_view_select` doesn't get a chance to work with them.

**Comment by ggevay** [comment]
Btw. the reason I looked into this is because I'm about to change our window function planning logic to be similar to aggregate functions (approach 4. or 5. [here](https://github.com/MaterializeInc/materialize/issues/20746)), and I was just about to create the exact same bug for window functions that we have here for aggregates...

**Comment by ggevay** [comment]
It would be good to quickly decide whether we want to fix this in a way that aggregates will be supported in ORDER BYs of set expressions or if we just want to improve the error as shown above, because this will inform how I will handle window functions while fixing https://github.com/MaterializeInc/materialize/issues/20746.

**Comment by mjibson** [comment]
> if we are ok with following Postgres in this, then the fix is as simple as flipping allow_aggregates to false

Yes, if we can fix this and improve postgres compat, we should do that.

Should we also switch `allow_windows: true` to `false` there too?

Does postgres allow window functions in union order by? If we need to disable those, we have a safe process for doing that. We can produce some code that would reject those, and then run our `stash-debug upgrade-check` tool to see if anything would fail. That avoids needing to write a query that can correctly figure this out. I'm in favor of disabling these if they are buggy or overly difficult.

**Comment by mjibson** [comment]
This seems like a thing sqlsmith should have found! Wonder why it didn't. cc @def-

**Comment by ggevay** [comment]
> Should we also switch allow_windows: true to false there too?
>
> Does postgres allow window functions in union order by? 
> If we need to disable those, we have a safe process for doing that. We can produce some code that would reject those, and then run our stash-debug upgrade-check tool to see if anything would fail. That avoids needing to write a query that can correctly figure this out. I'm in favor of disabling these if they are buggy or overly difficult.

Oh, that's great, thanks! I didn't know about that tool. Ok, then I'd say let's also switch `allow_windows` to false. Postgres also doesn't allow this. In fact, it only allows result column names, not even any expressions or functions. So even with aggregations and window functions disabled, we will still be better than Postgres.

Then I'll open a PR that turns `allow_aggregates` and `allow_windows` to false and adds some tests with various things in such ORDER BYs.

How does `stash-debug upgrade-check` work? Do I merge the PR before we can run the tool? Or do we run it while the PR is open? (Btw. does this tool run at every release during the release validation process?)

**Comment by mjibson** [comment]
> How does stash-debug upgrade-check work? Do I merge the PR before we can run the tool? Or do we run it while the PR is open?

It does not run automatically anywhere, even as part of the upgrade process. We have to explicitly request it happen as part of an upgrade. See https://github.com/MaterializeInc/cloud/issues/5057 for more.
